src/goto-programs/character_refine_preprocess.cpp

```diff
@@ -0,0 +1,2564 @@
+/*******************************************************************\
+
+Module: Preprocess a goto-programs so that calls to the java Character
+        library are replaced by simple expressions.
+
+Author: Romain Brenguier
+
+Date:   March 2017
+
+\*******************************************************************/
+
+#include <util/arith_tools.h>
+#include "character_refine_preprocess.h"
+
+/*******************************************************************\
+
+Function: character_refine_preprocesst::convert_char_function
+
+  Inputs:
+    expr_function - A reference to a function on expressions
+    target - A position in a goto program
+
+ Purpose: converts based on a function on expressions
+
+\*******************************************************************/
+
+void character_refine_preprocesst::convert_char_function(
+  exprt (*expr_function)(const exprt &chr, const typet &type),
+  conversion_input &target)
+{
+  const code_function_callt &function_call=to_code_function_call(target->code);
+  assert(function_call.arguments().size()==1);
+  exprt arg=function_call.arguments()[0];
+  exprt result=function_call.lhs();
+  target->make_assignment();
+  typet type=result.type();
```

Comment:
`const typet &type=result.type();`  If you delay the `target->make_assignment()` call until after `code` is constructed, you could even make all of the above `const exprt &`